Improve deeplinks experience

jpcloureiro opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.
-->

**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

_Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions,_
_1. What is the reason for the change?_
_2. What is the improvement/solution?_

**Screenshots/Recordings**

_If applicable, add screenshots and/or recordings to visualize the before and after of your change_

**Issue**

Progresses [#460](https://github.com/MetaMask/mobile-planning/issues/460)

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

joaoloureirop committed:
Browser: prompt user before deeplink execution

joaoloureirop committed:
Browser: translate protocol alert messages

NicholasEllul commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -1339,6 +1410,15 @@ export const BrowserTab = (props) => {
    */
+
+  const originWhitelist = Object.keys(knownProtocols).map((key) => {
+    const protocol = key.startsWith('http')
```
Do we know if theres a case that `http` could have any upper cases (ex: `hTTp://google.com`, or does that get normalized somewhere?

NicholasEllul commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -785,15 +785,86 @@ export const BrowserTab = (props) => {
+
+    const { alertMsg } = knownProtocols[protocol];
+
+    if (!isTopFrame && Boolean(alertMsg)) {
+      Alert.alert(alertMsg, '', [
+        {
+          text: 'Ignore',
+          onPress: () => null,
+          style: 'cancel',
+        },
+        {
+          text: 'Allow',
+          onPress: () => allowLinkOpen(url),
+          style: 'allow',
+        },
+      ]);
+      return false;
+    }
+
+    return true;
```
In the event a protocol comes in such as `foobar://` this if-statement will evaluate false as there is no alert message, and as a result the `verifyKnownProtocols` function and the `onShouldLoadWithRequest` will return true. This will lead to us proceeding with the execution of unknown protocols. Can we also have a generic alert to protect users in that case?

jpcloureiro replied:
I've modified the code so any non whitelisted protocol will trigger an alert dialog with a generic message

NEllusion reviewed (commented):
I love this approach of intercepting the request! Are we able to add any tests to catch regressions in this behaviour?

sethkfman commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -785,15 +785,86 @@ export const BrowserTab = (props) => {
+    if (!isTopFrame && Boolean(alertMsg)) {
+      Alert.alert(alertMsg, '', [
+        {
+          text: 'Ignore',
```
Check to see if there is something already available in our locales for translations.

andreahaku commented:
@jpcloureiro did you check how this change behaves with the app specific schemas like `metamask://`, `dapp://`, 'wc://' or 'ethereum://'?

joaoloureirop committed:
code review 1

jpcloureiro commented:
> @jpcloureiro did you check how this change behaves with the app specific schemas like `metamask://`, `dapp://`, `wc://` or `ethereum://`?

@andreahaku opening any of these schemas inside the webview will open the alert dialog. If the user presses allow, it will execute as before.

joaoloureirop committed:
remove placeholder unit test

joaoloureirop committed:
destructure url on onShouldStartLoadWithRequest args

sethkfman commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -785,6 +785,56 @@ export const BrowserTab = (props) => {
     });
   }, []);
 
+  const protocolWhitelist = ['about:', 'http:', 'https:'];
```
If any these are not impacting state, could they be moved out to a `ults` class that is TSX and can be tested?

joaoloureirop committed:
refactor code & add tests

move functions around where they best fit

joaoloureirop committed:
add protocolWhitelist tests

joaoloureirop committed:
fix lint errors

joaoloureirop committed:
browserUtils: code refactor

delete duplicate code & improve test coverage

joaoloureirop committed:
add JSDOC comments to new functions

joaoloureirop committed:
fix test description

sethkfman commented on app/util/browser/index.test.ts:
```diff
@@ -1,4 +1,29 @@
+} from '.';
+import { strings } from '../../../locales/i18n';
+
+describe('Browser utils :: sanitizeUrl', () => {
```
Nice work on these test cases

sethkfman commented on app/util/browser/index.ts:
```diff
@@ -0,0 +1,141 @@
+ * @param defaultProtocol - Protocol string to append to URLs that have none
+ * @returns - String corresponding to sanitized input depending if it's a search or url
+ */
+export const sanitizeUrl = (url: string, defaultProtocol = 'https://') => {
```
I like the effort to refactor to type script.

sethkfman commented on app/components/Views/BrowserTab/index.js:
```diff
@@ -784,14 +791,38 @@ export const BrowserTab = (props) => {
+
+    // Continue request loading it the protocol is whitelisted
+    const { protocol } = new URL(url);
+    if (protocolWhitelist.includes(protocol)) return true;
```
Could we call change this to `protocolAllowlist()`?

jpcloureiro replied:
Yes, changing `protocolWhitelist` to `protocolAllowList`

sethkfman reviewed (approved):
One minor comment for a method rename but other than that it looks good.

joaoloureirop committed:
rename protocolWhitelist to protocolAllowList

sethkfman reviewed (approved):
LGTM

NicholasEllul commented on app/util/browser/index.ts:
```diff
@@ -0,0 +1,141 @@
+      // In case of keywords we default to google search
+      let searchUrl = 'https://www.google.com/search?q=' + escape(input);
+      if (searchEngine === 'DuckDuckGo') {
+        searchUrl = 'https://duckduckgo.com/?q=' + escape(input);
```
Escape is now [deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape) and instead its recommended that we use [encodeUriComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)

jpcloureiro commented on app/util/browser/index.ts:
```diff
@@ -0,0 +1,141 @@
+ * @param defaultProtocol - Protocol string to append to URLs that have none
+ * @returns - String corresponding to sanitized input depending if it's a search or url
+ */
+export const sanitizeUrl = (url: string, defaultProtocol = 'https://') => {
+  const hasProtocol = /^[a-z]*:\/\//.test(url);
+  const sanitizedURL = hasProtocol ? url : `${defaultProtocol}${url}`;
+  return sanitizedURL;
+};
```
Agree, `prefixUrlWithProtocol` is better suited for what the function does.

joaoloureirop committed:
revision

- rename function sanitizeUrl to prefixUrlWithProtocol.
- replace deprecated escape with encodeURIComponent.

sethkfman commented:
@jpcloureiro Can you add the scenarios you tested and update the lint errors?

joaoloureirop committed:
(browser utils) update unit tests & lint

cortisiko reviewed (changes requested):
@jpcloureiro sending this back your way!

**Issue** 

On the webview error screen, when I tap the "return to home page" button I get:`[TypeError: Invalid URL: home.metamask.io]`. Furthermore, I am not taken back to the mm home page.  
see [recording](http://recordit.co/RncotziS6t)

To reproduce:
- open the browser
- enter an invalid URL. For example https://asfasfo.aox
- When the error screen appears
- tap on "return to home page". Notice the error.

joaoloureirop committed:
fix homepage url protocol prefix

joaoloureirop committed:
fix lint

jpcloureiro commented:
> @jpcloureiro sending this back your way!
> 
> **Issue**
> 
> On the webview error screen, when I tap the "return to home page" button I get:`[TypeError: Invalid URL: home.metamask.io]`. Furthermore, I am not taken back to the mm home page. see [recording](http://recordit.co/RncotziS6t)
> 
> To reproduce:
> 
> * open the browser
> * enter an invalid URL. For example https://asfasfo.aox
> * When the error screen appears
> * tap on "return to home page". Notice the error.

@cortisiko pushed a fix. It should go back to mm home page now.

cortisiko reviewed (approved):
Fix looks good @jpcloureiro, This is QA passed! 🌮 🌮
